base/src/main/java/io/spine/validate/option/RangeConstraint.java

```diff
@@ -21,99 +21,112 @@
+     *     <li>{@code [0 . . 1]} - divider cannot be split with spaces.
+     *     <li>{@code ( .. 0)} - missing lower bound.
+     * </ul>
+     */
+    private static final Pattern NUMBER_RANGE =
+            Pattern.compile("([\\[(])\\s*([+\\-]?[\\d.]+)\\s*\\.\\.\\s*([+\\-]?[\\d.]+)\\s*([])])");
 
-    RangeConstraint(String optionValue) {
-        super(optionValue, rangeFromOption(optionValue));
+    RangeConstraint(String optionValue, FieldDeclaration field) {
+        super(optionValue, rangeFromOption(optionValue, field), field);
     }
 
     @VisibleForTesting
-    static Range<ComparableNumber> rangeFromOption(String value) {
-        String trimmed = value.trim();
-        RangeType range = RangeType.parse(trimmed);
-        EdgeValues edgeValues = edgeValues(withoutBraces(trimmed));
-        ComparableNumber left = edgeValues.leftEdge.toNumber();
-        ComparableNumber right = edgeValues.rightEdge.toNumber();
-        Range<ComparableNumber> result = range.create(left, right);
-        return result;
+    static Range<ComparableNumber> rangeFromOption(String rangeOption, FieldDeclaration field) {
+        return !rangeOption.isEmpty()
+               ? compileRange(rangeOption, field)
+               : Range.all();
     }
 
-    private static EdgeValues edgeValues(String value) {
-        String trimmed = value.trim();
-        ImmutableList<String> edges = ImmutableList.copyOf(RANGE_SPLITTER.split(trimmed));
-        String leftEdge = edges.get(0);
-        String rightEdge = edges.get(1);
-        return EdgeValues.of(leftEdge, rightEdge);
+    private static Range<ComparableNumber>
+    compileRange(String rangeOption, FieldDeclaration field) {
+        Matcher rangeMatcher = NUMBER_RANGE.matcher(rangeOption.trim());
+        checkState(rangeMatcher.matches(),
+                   "Malformed range `%s` on the field `%s`. " +
+                   "Must have a form of `[a..b]` " +
+                   "where `a` and `b` are valid literals of the type `%s`. " +
+                   "See doc of `(range)` for more details.",
+                   rangeOption, field, field.javaTypeName());
+        boolean minInclusive = rangeMatcher.group(1).equals("[");
+        ComparableNumber minValue = new NumberText(rangeMatcher.group(2)).toNumber();
+        ComparableNumber maxValue = new NumberText(rangeMatcher.group(3)).toNumber();
+        boolean maxInclusive = rangeMatcher.group(4).equals("]");
+        if (minInclusive) {
+            return maxInclusive ? closed(minValue, maxValue) : closedOpen(minValue, maxValue);
+        } else {
+            return maxInclusive ? openClosed(minValue, maxValue) : open(minValue, maxValue);
+        }
     }
 
-    private static String withoutBraces(String value) {
-        return value.substring(1, value.length() - 1);
+    @Override
+    protected String compileErrorMessage(Range<ComparableNumber> range) {
+        return format("The value of the field `%s` is out of range. Must be %s%s and %s%s.",
```

Comment:
I've updated the language here myself, removing the word “between” which was redundant, assuming that following text says “less than”. But there's one more thing about this error message. What _is_ the value, which is out of range?

I think, we need to tell it to make finding the problem easier. Can we do that?

Reply:
It is possible to add the field value to the message. `(min)` and `(max)` also don't provide the value in the message. However, in both cases (of `(range)` and `(min)`/`(max)`) provide the field value in the violation body.
I'd rather leave this behaviour as is. This gives the users an additional perk by making the error message non-sensitive, i.e. such that does not contain potentially sensitive data.